pkg/privilege/privileges/cache.go

```diff
@@ -521,147 +527,143 @@ func findUserAndAllRoles(all map[string]struct{}, roleGraph map[string]roleGraph
 	err := p.loadTable(ctx, sqlLoadRoleGraph, p.decodeRoleEdgesTable)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	// Including the user list and also their roles
-	extendedUserList := make(map[string]struct{}, len(userList))
-	for _, user := range userList {
-		extendedUserList[user] = struct{}{}
-	}
-	findUserAndAllRoles(extendedUserList, p.roleGraph)
-	// Re-generate the user list.
-	userList = userList[:0]
-	for user := range extendedUserList {
-		userList = append(userList, user)
-	}
+	// Including the user and also their roles
+	userList := findUserAndAllRoles(users, p.roleGraph)
 
-	err = p.loadTable(ctx, sqlLoadUserTable, p.decodeUserTableRow, userList...)
+	err = p.loadTable(ctx, addUserFilterCondition(sqlLoadUserTable, userList), p.decodeUserTableRow(userList))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	err = p.loadTable(ctx, sqlLoadGlobalPrivTable, p.decodeGlobalPrivTableRow, userList...)
+	err = p.loadTable(ctx, addUserFilterCondition(sqlLoadGlobalPrivTable, userList), p.decodeGlobalPrivTableRow(userList))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	err = p.loadTable(ctx, sqlLoadGlobalGrantsTable, p.decodeGlobalGrantsTableRow, userList...)
+	err = p.loadTable(ctx, addUserFilterCondition(sqlLoadGlobalGrantsTable, userList), p.decodeGlobalGrantsTableRow(userList))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	err = p.loadTable(ctx, sqlLoadDBTable, p.decodeDBTableRow, userList...)
+	err = p.loadTable(ctx, addUserFilterCondition(sqlLoadDBTable, userList), p.decodeDBTableRow(userList))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	err = p.loadTable(ctx, sqlLoadTablePrivTable, p.decodeTablesPrivTableRow, userList...)
+	err = p.loadTable(ctx, addUserFilterCondition(sqlLoadTablePrivTable, userList), p.decodeTablesPrivTableRow(userList))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	err = p.loadTable(ctx, sqlLoadDefaultRoles, p.decodeDefaultRoleTableRow, userList...)
+	err = p.loadTable(ctx, addUserFilterCondition(sqlLoadDefaultRoles, userList), p.decodeDefaultRoleTableRow(userList))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
-	err = p.loadTable(ctx, sqlLoadColumnsPrivTable, p.decodeColumnsPrivTableRow, userList...)
+	err = p.loadTable(ctx, addUserFilterCondition(sqlLoadColumnsPrivTable, userList), p.decodeColumnsPrivTableRow(userList))
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
 	return userList, nil
 }
 
-// merge construct a new MySQLPrivilege by merging the data of the two objects;.
-func (p *MySQLPrivilege) merge(diff *MySQLPrivilege, userList []string) *MySQLPrivilege {
+// merge construct a new MySQLPrivilege by merging the data of the two objects.
+func (p *MySQLPrivilege) merge(diff *MySQLPrivilege, userList map[string]struct{}) *MySQLPrivilege {
 	ret := newMySQLPrivilege()
 	user := p.user.Clone()
-	for _, u := range userList {
-		user.Delete(itemUser{username: u})
+	for u := range userList {
```

Comment:
The new way: merge = old +- userList data (changes) 
and userList data = apply userList to diff (it can be data of userList or data of all users)

The new way is more general and O(n) to userList instead of O(n) to diff... the diff can be get by `LoadAll`